File: iplan/views/sidebar/sidebar_project.py
from gi.repository import Gtk, GLib, Gdk, Adw

from iplan.db.models.project import Project
from iplan.db.models.task import Task
from iplan.db.operations.list import read_lists
from iplan.db.operations.task import update_task, find_new_task_position
from iplan.views.project.project_list_task import ProjectListTask


@Gtk.Template(resource_path="/ir/imansalmani/iplan/ui/sidebar/sidebar_project.ui")
class SidebarProject(Gtk.ListBoxRow):
    __gtype_name__ = "SidebarProject"
    project: Project = None
    name_label: Gtk.Label = Gtk.Template.Child()

    def __init__(self, project):
        super().__init__()
        self.project = project

        self.name_label.set_label(self.project.name)

        if project.archive:
            self.name_label.add_css_class("dim-label")

        drop_task_target = Gtk.DropTarget.new(ProjectListTask, Gdk.DragAction.MOVE)
        drop_task_target.set_preload(True)
        drop_task_target.connect("drop", self.drop_task_target_drop_cb)
        drop_task_target.connect("motion", self.drop_task_target_motion_cb)
        self.add_controller(drop_task_target)

    @Gtk.Template.Callback()
    def drag_prepare_cb(
        self, drag_source: Gtk.DragSource, x: float, y: float
    ) -> Gdk.ContentProvider:
        return Gdk.ContentProvider.new_for_value(self)

    @Gtk.Template.Callback()
    def drag_begin_cb(self, drag_source: Gtk.DragSource, drag: Gdk.Drag):
        self.get_parent().select_row(self)
        drag_icon = Gtk.DragIcon.get_for_drag(drag)
        drag_icon.props.child = Gtk.Label()
        drag.set_hotspot(0, 0)

    @Gtk.Template.Callback()
    def drag_cancel_cb(
        self, drag_source: Gtk.DragSource, drag: Gdk.Drag, reason
    ) -> bool:
        self.get_parent().get_parent().select_active_project()
        return False

    def drop_task_target_drop_cb(
        self, target: Gtk.DropTarget, source_task_row: ProjectListTask, x, y
    ) -> bool:
        source_task_row.task.project = self.project._id
        source_task_row.task._list = list(read_lists(self.project._id))[0]._id
        source_task_row.task.position = find_new_task_position(
            source_task_row.task._list
        )
        source_task_row.get_parent().remove(source_task_row)
        update_task(source_task_row.task, move_position=True)
        self.get_parent().get_parent().select_active_project()
        return True

    def drop_task_target_motion_cb(self, target: Gtk.DropTarget, x, y):
        source_task_row: ProjectListTask = target.get_value()
        if source_task_row.task.project == self.project._id:
            return 0
        self.get_parent().select_row(self)
        return Gdk.DragAction.MOVE

File: src/views/sidebar/sidebar_project.rs
use gtk::{gdk, glib, glib::once_cell::sync::Lazy, prelude::*, subclass::prelude::*};
use std::cell::RefCell;

use crate::db::models::Project;
use crate::db::operations::read_projects;
use crate::views::sidebar::SidebarProjects;

mod imp {
    use super::*;

    #[derive(Default, gtk::CompositeTemplate)]
    #[template(resource = "/ir/imansalmani/iplan/ui/sidebar/sidebar_project.ui")]
    pub struct SidebarProject {
        pub project: RefCell<Project>,
        #[template_child]
        pub name_label: TemplateChild<gtk::Label>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for SidebarProject {
        const NAME: &'static str = "SidebarProject";
        type Type = super::SidebarProject;
        type ParentType = gtk::ListBoxRow;

        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
            klass.bind_template_instance_callbacks();
        }

        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for SidebarProject {
        fn constructed(&self) {
            self.parent_constructed();
        }

        fn properties() -> &'static [glib::ParamSpec] {
            static PROPERTIES: Lazy<Vec<glib::ParamSpec>> =
                Lazy::new(|| vec![glib::ParamSpecObject::builder::<Project>("project").build()]);
            PROPERTIES.as_ref()
        }

        fn set_property(&self, _id: usize, value: &glib::Value, pspec: &glib::ParamSpec) {
            match pspec.name() {
                "project" => {
                    let value = value.get::<Project>().expect("value must be a Project");
                    self.project.replace(value);
                }
                _ => unimplemented!(),
            }
        }

        fn property(&self, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
            match pspec.name() {
                "project" => self.project.borrow().to_value(),
                _ => unimplemented!(),
            }
        }
    }
    impl WidgetImpl for SidebarProject {}
    impl ListBoxRowImpl for SidebarProject {}
}

glib::wrapper! {
    pub struct SidebarProject(ObjectSubclass<imp::SidebarProject>)
        @extends gtk::Widget, gtk::ListBoxRow,
        @implements gtk::Buildable;
}

#[gtk::template_callbacks]
impl SidebarProject {
    pub fn new(project: Project) -> Self {
        let obj = glib::Object::new::<SidebarProject>(&[("project", &project)]);

        let imp = obj.imp();

        imp.name_label.set_text(&obj.project().name());

        if obj.project().archive() {
            imp.name_label.add_css_class("dim-label")
        };

        obj
    }

    pub fn project(&self) -> Project {
        self.property("project")
    }

    #[template_callback]
    fn handle_drag_prepare(&self, _x: f64, _y: f64) -> Option<gdk::ContentProvider> {
        Some(gdk::ContentProvider::for_value(&self.to_value()))
    }

    #[template_callback]
    fn handle_drag_begin(&self, drag: gdk::Drag) {
        self.parent()
            .unwrap()
            .downcast::<gtk::ListBox>()
            .unwrap()
            .select_row(Some(self));
        let drag_icon: gtk::DragIcon = gtk::DragIcon::for_drag(&drag).downcast().unwrap();
        let label = gtk::Label::builder().label("").build();
        drag_icon.set_child(Some(&label));
        drag.set_hotspot(0, 0);
    }

    #[template_callback]
    fn handle_drag_cancel(&self, _drag: gdk::Drag) -> bool {
        let projects_box = self.parent().and_downcast::<gtk::ListBox>().unwrap();
        let projects = read_projects(true).expect("Failed to read projects");
        let rows = projects_box.observe_children();
        for row in rows.into_iter() {
            let row: SidebarProject = row.unwrap().downcast().unwrap();
            let row_project = row.project();
            for project in &projects {
                if project.id() == row_project.id() {
                    row_project.set_property("index", project.index());
                    row.set_property("project", row_project);
                    break;
                }
            }
        }
        projects_box.invalidate_sort();
        projects_box.parent()
            .and_downcast::<SidebarProjects>()
            .unwrap()
            .select_active_project();
        false
    }
}

